Source/Icebreaker/Services/MatchingService.cs

```diff
@@ -204,9 +255,40 @@ private string GetChannelUserObjectId(ChannelAccount account)
             this.Randomize(users);
 
             var pairs = new List<Tuple<ChannelAccount, ChannelAccount>>();
-            for (int i = 0; i < users.Count - 1; i += 2)
+
+            var currentPairs = new HashSet<string>();
+
+            for (int i = 0; i < users.Count - 1; i++)
             {
-                pairs.Add(new Tuple<ChannelAccount, ChannelAccount>(users[i], users[i + 1]));
+                for (int j = i + 1; j < users.Count; j++)
+                {
+                    if (currentPairs.Contains(users[i].Id) || currentPairs.Contains(users[j].Id))
+                    {
+                        continue;
+                    }
+
+                    if (!pastPairs.ContainsKey(users[i].Id) || !pastPairs[users[i].Id].Contains(users[j].Id))
+                    {
+                        // match them
+                        pairs.Add(new Tuple<ChannelAccount, ChannelAccount>(users[i], users[j]));
+
+                        if (!pastPairs.ContainsKey(users[i].Id))
+                        {
+                            pastPairs[users[i].Id] = new HashSet<string>();
+                        }
+
+                        if (!pastPairs.ContainsKey(users[j].Id))
+                        {
+                            pastPairs[users[j].Id] = new HashSet<string>();
+                        }
+
+                        pastPairs[users[i].Id].Add(users[j].Id);
+                        pastPairs[users[j].Id].Add(users[i].Id);
+                        currentPairs.Add(users[i].Id);
+                        currentPairs.Add(users[j].Id);
+                        break;
+                    }
+                }
```

Comment:
Let's log if we couldn't find a match for a person

Reply:
Good idea. Perhaps we can add that in as a separate feature later on?